pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -623,338 +671,284 @@ func (r *Reconciler) shouldCreateNewResourceSnapshotNow(ctx context.Context, lat
+		klog.ErrorS(err, "Failed to parse the CRPGeneration from the annotations", "policySnapshot", latestKObj)
 		return controller.NewUnexpectedBehaviorError(err)
 	}
-	if crpGeneration != crp.Generation {
-		latest.Annotations[fleetv1beta1.CRPGenerationAnnotation] = strconv.FormatInt(crp.Generation, 10)
+	if crpGeneration != placementObj.GetGeneration() {
+		annotations := latest.GetAnnotations()
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
+		annotations[fleetv1beta1.CRPGenerationAnnotation] = strconv.FormatInt(placementObj.GetGeneration(), 10)
+		latest.SetAnnotations(annotations)
 		needUpdate = true
 	}
 
-	if crp.Spec.Policy != nil &&
-		crp.Spec.Policy.PlacementType == fleetv1beta1.PickNPlacementType &&
-		crp.Spec.Policy.NumberOfClusters != nil {
+	// Handle NumberOfClusters annotation for selectN type placements
+	placementSpec := placementObj.GetPlacementSpec()
+	if placementSpec.Policy != nil &&
+		placementSpec.Policy.PlacementType == fleetv1beta1.PickNPlacementType &&
+		placementSpec.Policy.NumberOfClusters != nil {
 		oldCount, err := annotations.ExtractNumOfClustersFromPolicySnapshot(latest)
 		if err != nil {
-			klog.ErrorS(err, "Failed to parse the numberOfClusterAnnotation", "clusterSchedulingPolicySnapshot", latestKObj)
+			klog.ErrorS(err, "Failed to parse the numberOfClusterAnnotation", "policySnapshot", latestKObj)
 			return controller.NewUnexpectedBehaviorError(err)
 		}
-		newCount := int(*crp.Spec.Policy.NumberOfClusters)
+		newCount := int(*placementSpec.Policy.NumberOfClusters)
 		if oldCount != newCount {
-			latest.Annotations[fleetv1beta1.NumberOfClustersAnnotation] = strconv.Itoa(newCount)
+			annotations := latest.GetAnnotations()
+			annotations[fleetv1beta1.NumberOfClustersAnnotation] = strconv.Itoa(newCount)
+			latest.SetAnnotations(annotations)
 			needUpdate = true
 		}
 	}
 	if !needUpdate {
 		return nil
 	}
 	if err := r.Client.Update(ctx, latest); err != nil {
-		klog.ErrorS(err, "Failed to update the clusterSchedulingPolicySnapshot", "clusterSchedulingPolicySnapshot", latestKObj)
+		klog.ErrorS(err, "Failed to update the policySnapshot", "policySnapshot", latestKObj)
 		return controller.NewUpdateIgnoreConflictError(err)
 	}
 	return nil
 }
 
-// ensureLatestResourceSnapshot ensures the latest resourceSnapshot has the isLatest label.
-func (r *Reconciler) ensureLatestResourceSnapshot(ctx context.Context, latest *fleetv1beta1.ClusterResourceSnapshot) error {
-	if latest.Labels[fleetv1beta1.IsLatestSnapshotLabel] == strconv.FormatBool(true) {
+// ensureLatestResourceSnapshot ensures the latest resourceSnapshot has the isLatest label, working with interface types.
+func (r *Reconciler) ensureLatestResourceSnapshot(ctx context.Context, latest fleetv1beta1.ResourceSnapshotObj) error {
+	labels := latest.GetLabels()
+	if labels != nil && labels[fleetv1beta1.IsLatestSnapshotLabel] == strconv.FormatBool(true) {
```

Comment:
no need to check if labels != nil,

when map is nil, it returns empty string and won't panic, 

https://go.dev/play/p/4Bv37T7kjlj